### PR TITLE
[MNG-6720] - Wait for parallel projects when using --fail-fast

### PR DESCRIPTION
**Pull request by oehme**
Otherwise the sessionFinished event may be sent before some
projects/goals finish, which violates the contracts of execution
listeners and makes it hard to properly lifecycle components that
are needed for goal execution.

Following this checklist to help us incorporate your 
contribution quickly and easily:

 - [x] Make sure there is a [JIRA issue](https://issues.apache.org/jira/browse/MNG) filed 
       for the change (usually before you start working on it).  Trivial changes like typos do not 
       require a JIRA issue.  Your pull request should address just this issue, without 
       pulling in other changes.
 - [x] Each commit in the pull request should have a meaningful subject line and body.
 - [x] Format the pull request title like `[MNG-XXX] - Fixes bug in ApproximateQuantiles`,
       where you replace `MNG-XXX` with the appropriate JIRA issue. Best practice
       is to use the JIRA issue title in the pull request title and in the first line of the 
       commit message.
 - [x] Write a pull request description that is detailed enough to understand what the pull request does, how, and why.
 - [x] Run `mvn clean verify` to make sure basic checks pass. A more thorough check will 
       be performed on your pull request automatically.
 - [x] You have run the [Core IT][core-its] successfully.

If your pull request is about ~20 lines of code you don't need to sign an
[Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf) if you are unsure
please ask on the developers list.

To make clear that you license your contribution under 
the [Apache License Version 2.0, January 2004](http://www.apache.org/licenses/LICENSE-2.0)
you have to acknowledge this by using the following check-box.

 - [x] I hereby declare this contribution to be licenced under the [Apache License Version 2.0, January 2004](http://www.apache.org/licenses/LICENSE-2.0)

 - [x] In any other case, please file an [Apache Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

[core-its]: https://maven.apache.org/core-its/core-it-suite/

### CODE REVIEW

**Comment by oehme** [comment]
Thanks!

**Comment by michael-o** [comment]
@oehme, keep contributing! +1

**Comment by Tibor17** [comment]
@oehme 
@michael-o 
@olamy 
Our build fails on assertion failure. Pls can you have a look and provide a fix? Thx
https://builds.apache.org/job/maven-box/job/maven/job/master/255/testReport/junit/org.apache.maven.it/MavenITmng6720FailFastTest/windows_jdk12___Run_ITs_Windows_Java_12___testItShouldWaitForConcurrentModulesToFinish/

```
mng6720FailFast(ItShouldWaitForConcurrentModulesToFinish)...FAILURE (0.9 s)
MavenITmng6720FailFastTest.java:67
```

**Comment by michael-o** [comment]
@Tibor17 I am confused how this can fail if the activation version is `[3.6.2,)` but we have 3.6.2-SNAPSHOT. No issue, I will check today.

**Comment by michael-o** [comment]
@Tibor17 I cannot reproduce it:

```
osipovmi@deblndw011x:~/var/Projekte/maven-integration-testing (master *=)
$ git log --format=%H | head -1
1d5db0a67c4595b853a012ec984c246ba078b374
[INFO] -------------------------------------------------------
[INFO]  T E S T S
[INFO] -------------------------------------------------------
[INFO] Running org.apache.maven.it.IntegrationTestSuite
Running integration tests for Maven 3.6.2-SNAPSHOT
        using Maven executable: /net/home/osipovmi/apache-maven-3.6.2-SNAPSHOT/bin/mvn
In order to disable '.mavenrc' set the system property 'maven.skip.rc', i.e. 'mvn -Dmaven.skip.rc=true -P run-its verify'.
Bootstrap(Bootstrap)........................................OK (4.6 s)
mng6720FailFast(ItShouldWaitForConcurrentModulesToFinish)...OK (4.0 s)
mng6506PackageAnnotation(GetPackageAnnotation)..............OK (1.6 s)
mng6256SpecialCharsAlternatePOMLocation(FolderWithSpace)....OK (0.2 s)
$ /net/home/osipovmi/apache-maven-3.6.2-SNAPSHOT/bin/mvn -v
Apache Maven 3.6.2-SNAPSHOT (93c88a2bd05bd1f318ad1174c2b3accdbe0cb36d; 2019-07-30T15:05:30+02:00)
Maven home: /net/home/osipovmi/apache-maven-3.6.2-SNAPSHOT
Java version: 1.8.0_212, vendor: Oracle Corporation, runtime: /usr/local/openjdk8/jre
Default locale: de_DE, platform encoding: UTF-8
OS name: "freebsd", version: "12.0-stable", arch: "amd64", family: "unix"
```

Can you reproduce it?

**Comment by michael-o** [comment]
Just have retried from my Windows desktop. Rebuilt Maven from master and ran ITs from master. All postiive. I cannot even access the Jenkins workspace on Windows to get `log.txt` for the IT.

**Comment by oehme** [comment]
According to the log, it already failed at the first assert, which is really weird - It would mean that not even the first project was executed. Maybe the log output paths are different on Jenkins somehow?

**Comment by michael-o** [comment]
@oehme That's the issue, we can't acces the log file :-(

**Comment by Tibor17** [comment]
Hey guys, I don't know root cause but I see what everybody can see on Jenkins.
It was just my notice for you ;-). You are the experts in this PR.

**Comment by oehme** [comment]
@michael-o let me know if I can help, but I'm afraid without access to those machines or the log I can't do much. It works well on my machine :/

**Comment by michael-o** [comment]
@oehme Same here :-(

**Comment by Tibor17** [comment]
@oehme 
Do you need to access the logs? If you do not have them ye, I can archive them in to Jenkins in ZIP and you may download them.

**Comment by michael-o** [comment]
@Tibor17 Please do so.

**Comment by Tibor17** [comment]
@oehme 
@michael-o 
The arvived ZIP files as well as the post-IT results including logs and Surefire reports is in `master`, see [1]. It is the first time. Before the archive was not here. So if you want to see the archives, pls trigger the build manually. I guess INFRA have switched the Git Hooks off. 
In your case the next build number will be 260. Simply download the ZIP files from https://builds.apache.org/job/maven-box/job/maven/job/master/260/artifact/ after the build 260 was triggered by you and the build has completed.

Let's hope that these archives help in the entire community. The key point is watching all the logs!

[1]: https://github.com/apache/maven/commit/cb71885c14eee51dc8a0f2031c19be3f358ec10c

**Comment by michael-o** [comment]
@Tibor17 Thanks, I have it already and am working on it. I understand the issue now and trying to fix piece by piece.

**Comment by oehme** [comment]
@michael-o Great! Can you summarize what the issue is? I'm curious :)

**Comment by michael-o** [comment]
@oehme This problem is not related to this change actually. There is a general problem of dependency resolution with a fresh repo depending on your environment. E.g., at work behind Nexus I need to add more dependencies to the bootstrap IT that at home. It was the same case for this one. See 12f31a47d8f60b2698031cf81031c1c51c9dd958, fee37c64b197ef094a1a966ddd60e2aaffbe612c, 8a4b7170eff20988c8ba1731dfa47a0c117d51f2,

**Comment by jira-importer** [comment]
Resolve #8458
